src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -82,23 +90,33 @@ class MongoDbStore implements StoreInterface
      *
      * @see https://docs.mongodb.com/manual/applications/replication/
      */
-    public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)
+    public function __construct($mongo, array $options = [], float $initialTtl = 300.0)
     {
-        $this->mongo = $mongo;
-
-        $this->initialTtl = $initialTtl;
-
         $this->options = array_merge([
-            'collection' => 'lock',
             'gcProbablity' => 0.001,
+            'collection' => 'lock',
+            'uriOptions' => [],
+            'driverOptions' => [],
         ], $options);
 
-        if (!isset($this->options['database'])) {
-            throw new InvalidArgumentException('You must provide the "database" option for MongoDBStore');
+        $this->initialTtl = $initialTtl;
+
+        if ($mongo instanceof Collection) {
+            $this->collection = $mongo;
+        } elseif ($mongo instanceof Client) {
+            $this->client = $mongo;
+        } elseif (is_string($mongo)) {
+            $this->dsn = $mongo;
+        } else {
+            throw new InvalidArgumentException(sprintf('%s() requires %s or %s or DSN as first argument, "%s" given.', __METHOD__, Collection::class, Client::class, \is_object($mongo) ? \get_class($mongo) : \gettype($mongo)));
+        }
+
+        if (null === $this->collection && !isset($this->options['database'])) {
+            throw new InvalidArgumentException(sprintf('%s() requires the "database" option when constructing with a %s', __METHOD__, \is_object($mongo) ? \get_class($mongo) : \gettype($mongo)));
```

Comment:
database and collection names could be defined in the DSN (similar to PDO). dsn = `mongodb://[username:password@]host1[:port1][,host2[:port2:],...]/[database/][collection]` ?

Reply:
The database you see there is the authentication database, not the database to use for data

Reply:
isn't it the purpose of the querystring parameter `authSource`? https://symfony.com/doc/master/bundles/DoctrineMongoDBBundle/installation.html#authentication

anyway, if the `path` component is a standard to declare the auth Database, we can use custom queryString parameters like `mongodb://localhost/?lockDatabase=foo&lockCollection=bar`

If we can not create a the Store from a DSN, developpers using the FrameworkBundle won't be able to use it from the configuration `framework.lock = '%env(LOCK_DNS)%` (see https://github.com/symfony/symfony-docs/pull/12523)

Reply:
> isn't it the purpose of the querystring parameter `authSource`?

Actually I just tested this (mongodb 4.2.0) and if both `/path` and `?authSource` are specified `/path` means application database and `?authSource` means authentication database. We are still stuck for getting the collection name.

>  we can use custom queryString parameters

Seems to be the only way, do you have any objection to us taking `?collection=bar`? I feel like if we ever had a conflict it would be because MongoDB is using this param for the exact same purpose.

I'm going to enforce when `$mongo` is a:

* `MongoDB\Collection`:
  * `$options['database']` is ignored
  * `$options['collection']` is ignored
* `MongoDB\Client`:
  * `$options['database']` is mandatory
  * `$options['collection']` is mandatory
* MongoDB Connection URI string:
  * `/path` is used otherwise `$options['database']`, at least one is mandatory
  * `?collection=` is used otherwise `$options['collection']`, at least one is mandatory